benchmarktests/target_auth_gcp.go

```diff
@@ -0,0 +1,284 @@
+	}
+
+	return nil
+}
+
+func (g *GCPAuth) Target(client *api.Client) vegeta.Target {
+	return vegeta.Target{
+		Method: GCPAuthTestMethod,
+		URL:    client.Address() + g.pathPrefix + "/login",
+		Header: g.header,
+		Body:   []byte(fmt.Sprintf(`{"role": "%s", "jwt": "%s"}`, g.roleName, g.jwt)),
+	}
+}
+
+func (g *GCPAuth) Cleanup(client *api.Client) error {
+	g.logger.Trace(cleanupLogMessage(g.pathPrefix))
+	_, err := client.Logical().Delete(strings.Replace(g.pathPrefix, "/v1/", "/sys/", 1))
+	if err != nil {
+		return fmt.Errorf("error cleaning up mount: %v", err)
+	}
+	return nil
+}
+
+func (g *GCPAuth) GetTargetInfo() TargetInfo {
+	return TargetInfo{
+		method:     GCPAuthTestMethod,
+		pathPrefix: g.pathPrefix,
+	}
+}
+
+func (g *GCPAuth) Setup(client *api.Client, randomMountName bool, mountName string) (BenchmarkBuilder, error) {
+	var err error
+	authPath := mountName
+	config := g.config.Config
+	g.logger = targetLogger.Named(GCPAuthTestType)
+
+	if randomMountName {
+		authPath, err = uuid.GenerateUUID()
+		if err != nil {
+			return nil, fmt.Errorf("can't generate UUID for mount name: %v", err)
+		}
+	}
+
+	g.logger.Trace(mountLogMessage("auth", "gcp", authPath))
+	err = client.Sys().EnableAuthWithOptions(authPath, &api.EnableAuthOptions{
+		Type: "gcp",
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error enabling gcp: %v", err)
+	}
+	setupLogger := g.logger.Named(authPath)
+
+	// check if the provided argument should be read from file
+	creds := config.GCPAuthConfig.Credentials
+	if len(creds) > 0 && creds[0] == '@' {
+		var err error
+		contents, err := ioutil.ReadFile(creds[1:])
+		if err != nil {
+			return nil, fmt.Errorf("error reading file: %w", err)
+		}
+
+		config.GCPAuthConfig.Credentials = string(contents)
+	}
+
+	setupLogger.Trace(parsingConfigLogMessage("gcp auth"))
+	GCPAuthConfig, err := structToMap(config.GCPAuthConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing gcp auth config from struct: %v", err)
+	}
+
+	// Write GCP config
+	setupLogger.Trace(writingLogMessage("gcp auth config"))
+	_, err = client.Logical().Write("auth/"+authPath+"/config", GCPAuthConfig)
+
+	if err != nil {
+		return nil, fmt.Errorf("error writing gcp config: %v", err)
+	}
+
+	setupLogger.Trace(parsingConfigLogMessage("role"))
+	GCPRoleConfig, err := structToMap(config.GCPTestRoleConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing role config from struct: %v", err)
+	}
+
+	// Write GCP Role
+	setupLogger.Trace(writingLogMessage("role"), "name", config.GCPTestRoleConfig.Name)
+	_, err = client.Logical().Write("auth/"+authPath+"/role/"+config.GCPTestRoleConfig.Name, GCPRoleConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error writing gcp role: %v", err)
+	}
+
+	// TODO: handle more than one service account
+	jwt, err := getSignedJwt(g.config.Config.GCPTestRoleConfig.Name, config.GCPAuthConfig.Credentials, config.GCPTestRoleConfig.MaxJWTExp, config.GCPTestRoleConfig.BoundServiceAccounts[0])
```

Comment:
this seems weird to me. The role allow for multiple service accounts, but in our test we only use the first one? Should we allow the user to specify which service account they want tested?

Reply:
added randomness so that other service accounts can be used to sign in as well.